### DISCUSSION TITLE
Android build has become very slow and unstable

### DISCUSSION THREAD

**Discussion by burekas7** [codemagic-ci-cd/codemagic-docs#2589] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

Recently the Android build has become very slow.

The only change was a change in the JAVA version:
17, instead of the default version which is currently 11
(By the way, at some point I recommend setting 17 as the default)

But nothing else has changed.

So why has Android build become so slow and unstable (with strange errors sometimes, or sdk installation)

**Some of the last builds:**
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/66674d93e516f48986aad146
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6667451d796e2a2623577b19 (The worst, Forced cancellation)
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6665d239bab98ad219a1104c (Contains strange errors that wasn't recovered afterwards)
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/66674b2c1bf7e57add693ece (Forced cancellation)
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6665d55364ef0f2a578dd30a
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/666632b594f1b2ba0dc1129c

**Comment by dtrdic** [comment]
hi @burekas7 If you expand the logs for the Installing dependencies step, you'll notice an error message indicating that the current Gradle version you are using (8.1.1) is not supported, and a minimum Gradle version of 8.2 is required. Please update your Gradle version to 8.2 or higher. 
Also, Maven Central experienced some [outages](https://status.maven.org/) recently, which might also be contributing to the issue you're encountering

Additionally, the default Java version on the build machine is already set to version 17, starting from June 10. Check [this announcement ](https://github.com/orgs/codemagic-ci-cd/discussions/2567)

**Reply by burekas7** [reply]
@dtrdic 
1) But the Gradle version is 8.2.2 in my code. (android/build.gradle)
2) "Maven Central experienced some [outages](https://status.maven.org/)" This may explaine some of the issues here.
3) Regarding the JAVA, thanks, so I will remove this from the Workflow params.

Thanks.

**Reply by dtrdic** [reply]
hey @burekas7 If you've set it in android/build.gradle then you've set the gradle plugin version, but the Gradle version is set in the gradle-wrapper.properties file. Try checking the distributionUrl in `gradle-wrapper.properties` it should also have gradle-8.2 version
Check this gradle [compatibility matrix](https://developer.android.com/build/releases/gradle-plugin#updating-gradle) for reference

<img width="872" alt="image" src="https://github.com/codemagic-ci-cd/codemagic-docs/assets/38918425/4b910be2-b5ce-4c2a-ba3f-b84dec7fa456">

**Reply by burekas7** [reply]
@dtrdic 
Also there, the version is "..../gradle-8.2-bin.zip" (you zip file is **all**, I don't know if it's matter)

**Reply by dtrdic** [reply]
@burekas7, it looks like that even though your android/build.gradle file specifies Gradle version 8.2.2, the actual version being used is 8.1.1.  
Also the `-bin` distribution might lack some necessary components that are included in the `-all `distribution so this could be the reason.. But try also setting `distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.2-all.zip` in gradle-wrapper.properties

**Reply by burekas7** [reply]
@dtrdic 

Still.
Very bad experience with Android build, unstable, slow....

New builds:
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6672ed06235f17ce2039d2dd
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6672eb48753fe1853ea6c724
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6672e9ee8865b8d6fa163ceb

Even when it does succeed it takes much longer:
* https://codemagic.io/app/636d13c539a617000f89bf6e/build/6672f1b717e113710f9499d3

All were forcibly stopped, even though the minutes were still counting even though the build was stopped in the middle. 
I think that's not normal in my opinion when you make another attempt and another attempt because of a problem that is not yours and still count it.

**Reply by burekas7** [reply]
@dtrdic
It seems that updating the gradle files helped and improved the time a bit,
but it remains to be seen how stable it is.

https://docs.flutter.dev/release/breaking-changes/flutter-gradle-plugin-apply

**Comment by ArthurMassonBit** [comment]


**Reply by himesh-cm** [reply]
Hey @ArthurMassonBit, 

Could you please update the gradle files as mentioned below in the docs
https://docs.flutter.dev/release/breaking-changes/flutter-gradle-plugin-apply

If you still have issues, please consider opening a new thread with your build ID.

**Comment by burekas7** [comment]
@dtrdic  @himesh-cm 

There is still slowness during the build.
This time is regarding the "Installing dependencies".
Sometimes it's ok and sometimes it's taking few minutes.
Like in this build:
667ad151aef910282ef46e2f

(And also the iOS build is more then usual here)

In addition I'm still getting this error message even my grade configure with 8.2.
```
* What went wrong:
An exception occurred applying plugin request [id: 'com.android.application']
> Failed to apply plugin 'com.android.internal.version-check'.
   > Minimum supported Gradle version is 8.2. Current version is 8.1.1. If using the gradle wrapper, try editing the distributionUrl in /Users/builder/clone/android/gradle/wrapper/gradle-wrapper.properties to gradle-8.2-all.zip
```

Why is that?

This is my 'gradle-wrapper.properties' which is ok.
```
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
distributionUrl=https\://services.gradle.org/distributions/gradle-8.2-all.zip
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
```

**Reply by himesh-cm** [reply]
Hey @burekas7,
Do you have a Gradle wrapper in your repo (gradlew file)
if not, could you please add it and re-run the build?

**Reply by burekas7** [reply]
@himesh-cm 
Of course, I checked that before. 
The file is in version controlled with git etc.

**Comment by dtrdic** [comment]
hey @burekas7 just note that build minutes with Android, especially the ones that contain very heavy gradle tasks and take a long time to complete, can be improved considerably by caching `$HOME/.gradle/caches` path

![image](https://github.com/user-attachments/assets/43420248-c6be-4c12-abf1-9fff30f7a831)

**Comment by dtrdic** [comment]
I'm tentatively closing the discussion since we haven't heard back from you. You can open the discussion again by replying to this message or opening a new discussion.
